docs/developer-guide/firstanalysis.md

```diff
@@ -0,0 +1,144 @@
+  include Printable.PrintSimple (struct
+    type t' = t
+    let name = name
+    let show = show
+  end)
+  let hash = Hashtbl.hash
+
+  (* Here as important domain-specific implementations,
+   * there are some mistakes for you to fix... *)
+  let of_int i =
+    if i < Int64.zero then Zero
+    else if i > Int64.zero then Zero
+    else Zero
+
+  let gt x y = match x,y with
+  | Pos, Neg | Zero, Neg -> true
+  | _ -> false
+
+end
+
+module SL =
+struct
+  include Lattice.Flat (Signs) (Printable.DefaultNames)
+  let of_int i = `Lifted (Signs.of_int i)
+
+  let gt x y = match x, y with
+  | `Lifted x, `Lifted y -> Signs.gt x y
+  | _ -> false
+end
+```
+
+There are some flawed definitions above to get you started, but let us first tell the analysis to actually use this domain.
+For that, we change the similar lines that currently use the unit domain to now use a map from variables to the newly created sign domains.
+
+```
+module D = MapDomain.MapBot (Basetype.Variables) (SL)
+module G = Lattice.Unit
+module C = D
+```
+
+We will not care about the globals and the context, but it's convenient to have the same context as the domain itself.
+
+
+## Implementing the sign analysis
+
+The key part now is to define transfer functions for assignment. Here is a simple version that will suffice for our example. Replace the definition of `assign` in your file with the following:
+
+```
+let eval (d: D.t) (exp: exp): SL.t = match exp with
+| Const (CInt64 (i, _, _)) -> SL.of_int i
+| Lval (Var x, NoOffset) -> D.find x d
+| _ -> SL.top ()
+
+let assign ctx (lval:lval) (rval:exp) : D.t =
+  let d = ctx.local in
+  match lval with
+  | Var x, NoOffset -> D.add x (eval d rval) d
+  | _ -> D.top ()
+```
+
+We only handle assignments of the form `x = e` where the right-hand side `e` is itself either a constant of type integer or a simple variable. For this to work, the definition of `SL.of_int` needs to be corrected to pick an appropriate sign abstraction of integers. There is no need to implement the transfer functions for branching for this example; it only relies on lattice join operations to correctly take both paths into account.
+
+At this point, it may be useful to use Goblint's HTML output to [see the result](../user-guide/inspecting.md) of the analysis. This will also include Goblnt's base analysis, which is needed to deal with function calls, but it may be useful to turn off integer analysis with `--disable int.ana.def_exc`, so that only your implemented analysis plays a role.
+
+## Checking assertions
+
+With this in place, we should have sufficient information to tell Goblint that the assertion does hold. We need to answer assertion queries as follows:
+
+```
+let assert_holds (d: D.t) (e:exp) = match e with
+| BinOp (Gt, e1, e2, t) -> SL.gt (eval d e1) (eval d e2)
+| _ -> false
+
+let query ctx (type a) (q: a Queries.t): a Queries.result =
+  let open Queries in match q with
+  | Assert e when assert_holds ctx.local e -> `Lifted true
```

Comment:
I needed something that works now, but what are the plans once this is removed? Anyway, it's not clear what would be the best way to integrate the analysis into getting something that visibly does something. The student was a little bit overwhelmed by the types here, so maybe the query system isn't the best thing to involve in a "first tutorial" anyway.

Reply:
> I needed something that works now, but what are the plans once this is removed?

The plan is that for an assertion `EvalInt` query is made instead and one can answer whether it's true or false or unknown through its integer value, which it ultimately has in C anyway.

> Anyway, it's not clear what would be the best way to integrate the analysis into getting something that visibly does something. The student was a little bit overwhelmed by the types here, so maybe the query system isn't the best thing to involve in a "first tutorial" anyway.

Good point. Maybe avoid queries here indeed and just implement `special` to handle the `Assert` case to print the result out itself, e.g. "Assertion x < y holds in signs analysis". It also makes it clear that the assertion holds just due to the signs analysis, not due to something base did itself.

It would also remove the dependency on base analysis here, because whichever query is used, that depends on the base analysis doing the query for an assertion and printing the output, so it wouldn't be possible to run without base at all to see any output.

Reply:
But on the other hand, it is nice to showcase the query system and that these extra analyses can help improve stuff and not just work separately. Maybe, I just switch to the Int query and then if it really is a problem for students one could change it.

Reply:
The query system (and probably the event system too) might deserve separate documentation anyway, that also shows how to use `ctx.ask`, because using existing queries is probably more common than implementing new ones. Also, the explanation of the latter needs to go into some details like how may queries are different from must queries.